cloudevents/tests/test_kafka_conversions.py

```diff
@@ -0,0 +1,517 @@
+    for key, val in source_event.get_attributes().items():
+        assert event[key] == val
+    for key, val in source_event.data.items():
+        assert event.data[key] == val
+
+
+def test_to_structured_sets_value_default_marshallers(source_event):
+    result = to_structured(source_event)
+    assert result.value == json.dumps(
+        {
+            "specversion": source_event["specversion"],
+            "id": source_event["id"],
+            "source": source_event["source"],
+            "type": source_event["type"],
+            "time": source_event["time"],
+            "data": expected_data,
+        }
+    ).encode("utf-8")
+
+
+def test_to_structured_sets_value_custom_data_marshaller_default_envelope(
+    source_event, custom_marshaller
+):
+    result = to_structured(source_event, data_marshaller=custom_marshaller)
+    assert result.value == json.dumps(
+        {
+            "specversion": source_event["specversion"],
+            "id": source_event["id"],
+            "source": source_event["source"],
+            "type": source_event["type"],
+            "time": source_event["time"],
+            "data_base64": base64.b64encode(custom_marshaller(expected_data)).decode(
+                "ascii"
+            ),
+        }
+    ).encode("utf-8")
+
+
+def test_to_structured_sets_value_custom_envelope_marshaller(
+    source_event, custom_marshaller
+):
+    result = to_structured(source_event, envelope_marshaller=custom_marshaller)
+    assert result.value == custom_marshaller(
+        {
+            "specversion": source_event["specversion"],
+            "id": source_event["id"],
+            "source": source_event["source"],
+            "type": source_event["type"],
+            "time": source_event["time"],
+            "data": expected_data,
+        }
+    )
+
+
+def test_to_structured_sets_value_custom_marshallers(source_event, custom_marshaller):
+    result = to_structured(
+        source_event,
+        data_marshaller=custom_marshaller,
+        envelope_marshaller=custom_marshaller,
+    )
+    assert result.value == custom_marshaller(
+        {
+            "specversion": source_event["specversion"],
+            "id": source_event["id"],
+            "source": source_event["source"],
+            "type": source_event["type"],
+            "time": source_event["time"],
+            "data_base64": base64.b64encode(custom_marshaller(expected_data)).decode(
+                "ascii"
+            ),
+        }
+    )
+
+
+def test_to_structured_sets_key(source_event):
+    result = to_structured(source_event)
+    assert result.key == source_event["key"]
+
+
+def test_to_structured_none_key(source_event):
+    source_event["key"] = None
+    result = to_structured(source_event)
+    assert result.key is None
+
+
+def test_to_structured_no_key(source_event):
+    del source_event["key"]
+    result = to_structured(source_event)
+    assert result.key is None
```

Comment:
I believe these tests should actually assert that an error is raised when no `partitionkey` is available in the event or the mapping function fails to do it's job.